Clarify compatibility and content negotiation (#273)
OpenObservability/OpenMetrics: specification/OpenMetrics.md

Partial or invalid expositions MUST be considered erroneous in their entirety.

## Versioning and compatibility

### ABNF specification

> Versioning follows a semantic versioning model on the specification level. Breaking semantic changes (for example removal from ABNF) MUST result in major version bump, however additions MUST be in a minor version bump. Multiple changes may be bundled in the same version update and the version bump will be the biggest version bump in those changes.

Reviewer: ```suggestion
Versioning follows a semantic versioning model on the specification level. Breaking semantic changes MUST be signaled with a major version increase. For example, removing a stanza.

Semantic extensions to the ABNF MUST be signaled with a minor or major version increase. For exmaple, adding a line starting with `# foo`.
```

Reviewer: Semantic versioning needs a reference in line 38 and following, now.

Author: Added to informative references

### Ingestor

> Ingestors MUST implement at least version 1.0.0 of the standard.
> Ingestors MUST implement protocol negotiation to be able to support higher than 1.0.0 version.
> If exposer doesn't support content negotiation, version 1.0.0 SHOULD be assumed.

Reviewer: ```suggestion
Ingestors MUST implement version 1.0.0 of the standard.
Ingestors MAY support versions higher than 1.0.0. In this case, they MUST implement protocol version negotiation.
In case no protocol version negotiation takes place, version 1.0.0 SHOULD be assumed.
```

Reviewer: We can change the "MUST 1.0.0" to "SHOULD 1.0.0" in 2.x if we need to. It's a bit of a cop-out, but long term, say Prometheus 4.0, it might be defensible.

Author: Maybe, but I'm not sure we'll ever be able to get rid of 1.0.0

> An ingestor that supports version "m.y" of the standard MUST support versions "m.x" where "x" and "y" are natural numbers, such that "x < y".

Reviewer: I think supporting e.g. 1.0 and 1.5 would be OK.

Author: I'll remove this line

Author: I think we should state this explicitly, because that's not what backwards compatibility means. I would expect that a parser can read anything that lower semantic version (on the same major version).

> Ingestors MAY refuse to ingest the exposed metrics if content negotiation yields an empty list of versions.

Reviewer: That seems self-evident?

Reviewer: But they SHOULD negotiate the latest version as well?

Author: This is pretty academic, as negotiation currently takes place in the exposer, but the alternative would be to try and parse the content as 1.0.0.

I'll remove it as not very useful for now.

### Exposer

> Exposers MUST implement at least version 1.0.0 of the standard.
> Exposers MUST implement protocol negotiation to be able to support higher than 1.0.0 version.
> Exposers SHOULD negotiate the highest common version with the exposer.

Reviewer: ```suggestion
Exposers MUST implement version 1.0.0 of the standard.
Exposers MAY support versions higher than 1.0.0. In this case, they MUST implement protocol version negotiation.
Exposers SHOULD negotiate the highest common version with ingestors.
```
